master/master.go

```diff
@@ -197,43 +229,32 @@ func (ms *Master) distributeMapTask() {
 		workerID = (workerID + 1) % nWorkers
 	}
 
-	count := 0
-LOOP:
-	for {
-		select {
-		case <-finish:
-			count += 1
-			if count == len(ms.MapTasks) {
-				close(finish)
-				break LOOP
-			}
-		case crashUUID := <-ms.crashChan:
-			// ms.waitForIDLEWorkers()
-			workers, _ := ms.availableWorkers(1)
-			log.Info("[Master] Re-execute Map Task from ", crashUUID, " To ", workers[0].UUID)
-			if crashUUID == workers[0].UUID {
-				log.Panic("Self loop")
-			}
-			reExecuteTask, ok := taskStates.Load(crashUUID)
-			if !ok {
-				log.Panic("Load task states from crashUUID fail")
-			}
-			go func(task MapTaskInfo, id int) {
-				// taskStates.Delete(crashUUID)
-				taskStates.Store(workers[id].UUID, task)
-				done := Map(workers[id].IP, task.toRPC())
-				if !done {
-					task.setState(TASK_IDLE)
-					workers[id].SetState(WORKER_UNKNOWN)
-					ms.crashChan <- workers[id].UUID
-				} else {
-					task.setState(TASK_COMPLETED)
-					workers[id].SetState(WORKER_IDLE)
-					finish <- workers[id].UUID
-				}
-			}(reExecuteTask.(MapTaskInfo), 0)
+	for crashUUID := range orDone(finish, ms.crashChan, len(ms.MapTasks)) {
+		workers, _ := ms.availableWorkers(1)
+		log.Info("[Master] Re-execute Map Task from ", crashUUID, " To ", workers[0].UUID)
+		if crashUUID == workers[0].UUID {
+			log.Panic("Self loop")
 		}
+		reExecuteTask, ok := taskStates.Load(crashUUID)
+		if !ok {
+			log.Panic("Load task states from crashUUID fail")
+		}
+		go func(task MapTaskInfo, id int) {
+			// taskStates.Delete(crashUUID)
+			taskStates.Store(workers[id].UUID, task)
+			done := ms.client.Map(workers[id].IP, task.toRPC())
+			if !done {
+				task.setState(TASK_IDLE)
+				workers[id].SetState(WORKER_UNKNOWN)
+				ms.crashChan <- workers[id].UUID
+			} else {
+				task.setState(TASK_COMPLETED)
+				workers[id].SetState(WORKER_IDLE)
+				finish <- workers[id].UUID
+			}
+		}(reExecuteTask.(MapTaskInfo), 0)
 	}
+
```

Comment:
What if we end the executing process without crashing? This refactor didn't handle the normal end procedure.

Reply:
In this case orDone func return nothing (because we return in orDonefunc and defer close valStream will execute)and we do normal job